### DISCUSSION TITLE
Does Azure AD Authentication work?

### DISCUSSION THREAD

**Discussion by Viajaz** [requarks/wiki#4613] [Help / Questions]
Does the Azure AD Authentication module actually work in 2.5? Has someone gotten the Open ID Connect Authentication module working with Azure AD as a work around?
https://github.com/Requarks/wiki/discussions/3524

**Comment by Azayzel** [comment]
I've gotten both working. You need to make sure you're setting up your app registration as 'Web' with the correct redirect URL (found under the Azure Config Inputs). What error are you seeing?

**Reply by Viajaz** [reply]
The Wiki.js callback responds with HTTP error 500. No other information because Wiki.js doesn't have error logging for this feature.

Azure App registration includes an appropriate Redirect URI, `oauth2AllowIdTokenImplicitFlow` is true and `email`, `preferred_username` and `groups` claims are added to the tokens. I've granted the delegated scopes of `User.Read`, `email`, `openid` and `profile` in the Azure app.

**Reply by Viajaz** [reply]
_Identity Metadata Endpoint_ (`identityMetadata`) follows the appropriate `https://login.microsoftonline.com/ffffffff-ffff-ffff-ffff-ffffffffffff/v2.0/.well-known/openid-configuration` format as per [AzureAD/passport-azure-ad 4.1.1.2 Options](https://github.com/AzureAD/passport-azure-ad#4112-options)

I use 44x random ASCII characters as a _Cookie Encryption Key String_ (`cookieEncryptionKeyString`), there doesn't appear to be any specific required format as per [/server/modules/authentication/azure/authentication.js:21](https://github.com/Requarks/wiki/blob/latest/server/modules/authentication/azure/authentication.js#L21l)

**Redacted Azure AD Application Manifest**
```json
{
	"id": "000000000-0000-0000-0000-000000000000",
	"acceptMappedClaims": null,
	"accessTokenAcceptedVersion": null,
	"addIns": [],
	"allowPublicClient": false,
	"appId": "111111111-1111-1111-1111-111111111111",
	"appRoles": [],
	"oauth2AllowUrlPathMatching": false,
	"createdDateTime": "2021-10-21T00:00:00Z",
	"certification": null,
	"disabledByMicrosoftStatus": null,
	"groupMembershipClaims": "SecurityGroup",
	"identifierUris": [],
	"informationalUrls": {
		"termsOfService": null,
		"support": null,
		"privacy": null,
		"marketing": null
	},
	"keyCredentials": [],
	"knownClientApplications": [],
	"logoUrl": "https://redacted.com",
	"logoutUrl": null,
	"name": "Wiki.js",
	"oauth2AllowIdTokenImplicitFlow": true,
	"oauth2AllowImplicitFlow": false,
	"oauth2Permissions": [],
	"oauth2RequirePostResponse": false,
	"optionalClaims": {
		"idToken": [
			{
				"name": "email",
				"source": null,
				"essential": false,
				"additionalProperties": []
			},
			{
				"name": "groups",
				"source": null,
				"essential": false,
				"additionalProperties": [
					"sam_account_name"
				]
			},
			{
				"name": "preferred_username",
				"source": null,
				"essential": false,
				"additionalProperties": []
			}
		],
		"accessToken": [
			{
				"name": "email",
				"source": null,
				"essential": false,
				"additionalProperties": []
			},
			{
				"name": "groups",
				"source": null,
				"essential": false,
				"additionalProperties": [
					"sam_account_name"
				]
			},
			{
				"name": "preferred_username",
				"source": null,
				"essential": false,
				"additionalProperties": []
			}
		],
		"saml2Token": [
			{
				"name": "groups",
				"source": null,
				"essential": false,
				"additionalProperties": []
			}
		]
	},
	"orgRestrictions": [],
	"parentalControlSettings": {
		"countriesBlockedForMinors": [],
		"legalAgeGroupRule": "Allow"
	},
	"passwordCredentials": [
		{
			"customKeyIdentifier": null,
			"endDate": "2023-10-21T00:00:00Z",
			"keyId": "22222222-2222-2222-2222-222222222222",
			"startDate": "2021-10-21T00:00:00Z",
			"value": null,
			"createdOn": "2021-10-21T00:00:00Z",
			"hint": "FFF",
			"displayName": "Wiki.js"
		}
	],
	"preAuthorizedApplications": [],
	"publisherDomain": "redacted.com",
	"replyUrlsWithType": [
		{
			"url": "https://wiki.redacted.com/login/33333333-3333-3333-3333-333333333333/callback",
			"type": "Web"
		}
	],
	"requiredResourceAccess": [
		{
			"resourceAppId": "00000003-0000-0000-c000-000000000000",
			"resourceAccess": [
				{
					"id": "e1fe6dd8-ba31-4d61-89e7-88639da4683d",
					"type": "Scope"
				},
				{
					"id": "37f7f235-527c-4136-accd-4a02d197296e",
					"type": "Scope"
				},
				{
					"id": "64a6cdd6-aab1-4aaf-94b8-3cc8405e90d0",
					"type": "Scope"
				},
				{
					"id": "14dad69e-099b-42c9-810b-d002981feec1",
					"type": "Scope"
				}
			]
		}
	],
	"samlMetadataUrl": null,
	"signInUrl": "https://wiki.redacted.com",
	"signInAudience": "AzureADMyOrg",
	"tags": [],
	"tokenEncryptionKeyId": null
}
```

**Reply by Azayzel** [reply]
> _Identity Metadata Endpoint_ (`identityMetadata`) follows the appropriate `https://login.microsoftonline.com/ffffffff-ffff-ffff-ffff-ffffffffffff/v2.0/.well-known/openid-configuration` format as per [AzureAD/passport-azure-ad 4.1.1.2 Options](https://github.com/AzureAD/passport-azure-ad#4112-options)
> 
> 
> 
> I use 44x random ASCII characters as a _Cookie Encryption Key String_ (`cookieEncryptionKeyString`), there doesn't appear to be any specific required format as per [/server/modules/authentication/azure/authentication.js:21](https://github.com/Requarks/wiki/blob/latest/server/modules/authentication/azure/authentication.js#L21l)
> 
> 
> 
> **Redacted Azure AD Application Manifest**
> 
> ```json
> 
> {
> 
> 	"id": "000000000-0000-0000-0000-000000000000",
> 
> 	"acceptMappedClaims": null,
> 
> 	"accessTokenAcceptedVersion": null,
> 
> 	"addIns": [],
> 
> 	"allowPublicClient": false,
> 
> 	"appId": "111111111-1111-1111-1111-111111111111",
> 
> 	"appRoles": [],
> 
> 	"oauth2AllowUrlPathMatching": false,
> 
> 	"createdDateTime": "2021-10-21T00:00:00Z",
> 
> 	"certification": null,
> 
> 	"disabledByMicrosoftStatus": null,
> 
> 	"groupMembershipClaims": "SecurityGroup",
> 
> 	"identifierUris": [],
> 
> 	"informationalUrls": {
> 
> 		"termsOfService": null,
> 
> 		"support": null,
> 
> 		"privacy": null,
> 
> 		"marketing": null
> 
> 	},
> 
> 	"keyCredentials": [],
> 
> 	"knownClientApplications": [],
> 
> 	"logoUrl": "https://redacted.com",
> 
> 	"logoutUrl": null,
> 
> 	"name": "Wiki.js",
> 
> 	"oauth2AllowIdTokenImplicitFlow": true,
> 
> 	"oauth2AllowImplicitFlow": false,
> 
> 	"oauth2Permissions": [],
> 
> 	"oauth2RequirePostResponse": false,
> 
> 	"optionalClaims": {
> 
> 		"idToken": [
> 
> 			{
> 
> 				"name": "email",
> 
> 				"source": null,
> 
> 				"essential": false,
> 
> 				"additionalProperties": []
> 
> 			},
> 
> 			{
> 
> 				"name": "groups",
> 
> 				"source": null,
> 
> 				"essential": false,
> 
> 				"additionalProperties": [
> 
> 					"sam_account_name"
> 
> 				]
> 
> 			},
> 
> 			{
> 
> 				"name": "preferred_username",
> 
> 				"source": null,
> 
> 				"essential": false,
> 
> 				"additionalProperties": []
> 
> 			}
> 
> 		],
> 
> 		"accessToken": [
> 
> 			{
> 
> 				"name": "email",
> 
> 				"source": null,
> 
> 				"essential": false,
> 
> 				"additionalProperties": []
> 
> 			},
> 
> 			{
> 
> 				"name": "groups",
> 
> 				"source": null,
> 
> 				"essential": false,
> 
> 				"additionalProperties": [
> 
> 					"sam_account_name"
> 
> 				]
> 
> 			},
> 
> 			{
> 
> 				"name": "preferred_username",
> 
> 				"source": null,
> 
> 				"essential": false,
> 
> 				"additionalProperties": []
> 
> 			}
> 
> 		],
> 
> 		"saml2Token": [
> 
> 			{
> 
> 				"name": "groups",
> 
> 				"source": null,
> 
> 				"essential": false,
> 
> 				"additionalProperties": []
> 
> 			}
> 
> 		]
> 
> 	},
> 
> 	"orgRestrictions": [],
> 
> 	"parentalControlSettings": {
> 
> 		"countriesBlockedForMinors": [],
> 
> 		"legalAgeGroupRule": "Allow"
> 
> 	},
> 
> 	"passwordCredentials": [
> 
> 		{
> 
> 			"customKeyIdentifier": null,
> 
> 			"endDate": "2023-10-21T00:00:00Z",
> 
> 			"keyId": "22222222-2222-2222-2222-222222222222",
> 
> 			"startDate": "2021-10-21T00:00:00Z",
> 
> 			"value": null,
> 
> 			"createdOn": "2021-10-21T00:00:00Z",
> 
> 			"hint": "FFF",
> 
> 			"displayName": "Wiki.js"
> 
> 		}
> 
> 	],
> 
> 	"preAuthorizedApplications": [],
> 
> 	"publisherDomain": "redacted.com",
> 
> 	"replyUrlsWithType": [
> 
> 		{
> 
> 			"url": "https://wiki.redacted.com/login/33333333-3333-3333-3333-333333333333/callback",
> 
> 			"type": "Web"
> 
> 		}
> 
> 	],
> 
> 	"requiredResourceAccess": [
> 
> 		{
> 
> 			"resourceAppId": "00000003-0000-0000-c000-000000000000",
> 
> 			"resourceAccess": [
> 
> 				{
> 
> 					"id": "e1fe6dd8-ba31-4d61-89e7-88639da4683d",
> 
> 					"type": "Scope"
> 
> 				},
> 
> 				{
> 
> 					"id": "37f7f235-527c-4136-accd-4a02d197296e",
> 
> 					"type": "Scope"
> 
> 				},
> 
> 				{
> 
> 					"id": "64a6cdd6-aab1-4aaf-94b8-3cc8405e90d0",
> 
> 					"type": "Scope"
> 
> 				},
> 
> 				{
> 
> 					"id": "14dad69e-099b-42c9-810b-d002981feec1",
> 
> 					"type": "Scope"
> 
> 				}
> 
> 			]
> 
> 		}
> 
> 	],
> 
> 	"samlMetadataUrl": null,
> 
> 	"signInUrl": "https://wiki.redacted.com",
> 
> 	"signInAudience": "AzureADMyOrg",
> 
> 	"tags": [],
> 
> 	"tokenEncryptionKeyId": null
> 
> }
> 
> ```
> 
> 
> 
> 

Everything seems correct besides the secret. That secret should be coming from the Azure App Registration, not made up by you. You need to generate that secret in "Certificates and Secrets" under your App Registration.

**Reply by Viajaz** [reply]
@Azayzel That's just the Azure App Registration manifest straight from Azure with basic redactions, I've already generated a client secret as evident by the `keyId` under `passwordCredentials`.

For the Azure AD Authentication Strategy; If I disable the _Allow self-registration_ feature, login works.

**Reply by fandd1234** [reply]
@Viajaz I also got problems on using Azure AD authentication.  When I click on Azure AD to attempt for the login, it gave me "Oops, something went wrong...Invalid authentication provider".  May I ask if you know how to fix it please?  Thank you.

**Reply by Viajaz** [reply]
@fandd1234 I didn't fix it, I had to disable _Allow self-registration_ then I built an out-of-band batchjob script that registers in-scope users via the GraphQL API instead, it was tedious.

**Comment by JulianWeis** [comment]
@Viajaz @fandd1234  got the same problem using AAD any suggestions?
Thanks for your help

**Reply by Viajaz** [reply]
I didn't fix it, I had to disable Allow self-registration then I built an out-of-band batchjob script that registers in-scope users via the GraphQL API instead, it was tedious.

**Reply by loic667** [reply]
It used to work fine for us but since 2.5.255 it's broken (or maybe the version before that): can log in and out through AAD. Any idea how to solve this?

**Reply by RyanJ-MP** [reply]
> It used to work fine for us but since 2.5.255 it's broken (or maybe the version before that): can log in and out through AAD. Any idea how to solve this?

hi @loic667 did you have any luck with this? I am experiencing the same problem

**Reply by amal-viswa** [reply]
Hey, i'm having trouble while setting up azure ad authentication for wiki.js , I've followed the wiki.js official documentation for azure ad , but still not able to login error . can you guys help me resolve it

error -
![Screenshot 2022-08-23 at 18-12-26 Error](https://user-images.githubusercontent.com/90927394/186367130-4a6417c4-4b8b-446d-bf7e-2a6511e93203.png)

**Reply by RyanJ-MP** [reply]
@amal-viswa i think the code must be broken, I am getting the same as you. I have tried things others have suggested as a fix with no luck
https://github.com/requarks/wiki/discussions/5300

**Reply by loic667** [reply]
Hey @RyanJ-MP @amal-viswa , we had to keep using version 2.5.219 for AAD to work. Newer versions didn't work for us but haven't tested the most recent ones lately.

**Comment by rubentalstra** [comment]
@NGPixel my fix for the AzureAD authorization fixes all these problems with AzureAD (you where missing something, client Secret)

**Comment by eerotki** [comment]
I just set up Azure AD SSO with Wiki.js 2.5.289 according to [docs](https://docs.requarks.io/auth/azure) and got it working flawlessly.

I first had issues due to Microsoft Azure AD management interface. When in App Registrations in Azure AD, you copy the _OpenID Connect metadata document_ from Endpoints menu. The setup docs say there should be `YOUR_TENANT_ID` in the middle of the URL. **Make sure this Tenant ID is not "organizations" in your copied URL but a much longer string with letters, numbers and dashes!** For some reason the Endpoints menu first displayed an incorrect URL, and then another time it was displayed correctly with the correct Tenant ID. The Tenant ID should also be displayed in other endpoint URLs in the same menu (Federation metadata document, WS-Federation sign-on endpoint, ...) so **replace "organizations" part with the correct Tenant ID when configuring the SSO in Wiki.js.**

Also, make sure you configure Self-registration if you don't plan to add users in advance to Wiki.js. If you don't do this and you try to login with an user that has not been configured to Wiki.js, you get "You are not authorized" error during login.

**Reply by rubentalstra** [reply]
I know it works. But it’s better to setup with “client secret” so if you want to use Microsoft Graph Api you can use that to retrieve company information.

**Comment by patrick666rocker** [comment]
Greetings,

just finished Setting up 2.5.298 on fresh Ubuntu 22.04LTS and trying to get AzureAD Auth running - but running into
![image](https://user-images.githubusercontent.com/126105239/232863870-e844a330-a5c8-4ab9-9795-7bdcda4d2d91.png)

According to Azure, the Logins are Working, yet i still get the error message :(

**Reply by selcukkubur** [reply]
I have the same issue as well, any idea how we could fix ?

**Comment by LucasSPS** [comment]
After reading many topics about the same issue and testing several solutions, I finally reached a result that works. So, I’ll demonstrate how my environment (Wiki.js and Azure) is configured so you can reproduce it and fix the error. I’m not sure exactly which step solved it, but the combination led to a positive outcome.

My goal with the wiki was to ensure that all users, whether "guest" or "developers", had to be authenticated via Azure AD. So, as soon as the portal was opened, users would be redirected to the Azure AD login.

In Wiki.js, under the **Groups** tab, I removed all permissions and rules in **Page Rules** for the **Guests** group. This way, since nothing could be viewed, the redirect to login was in place.

For authentication, I set it up as follows:

* **clientId** is the same clientId that must be obtained from the Azure application.
* **Cookie Encryption Key** must be something random, like the one I inserted.

  <img width="1586" height="843" alt="image" src="https://github.com/user-attachments/assets/aae188f4-8d7d-4c7f-a063-98a351bfd9fe" />  

</br></br>

* Allowing **self-registration** is essential so new users can authenticate.

  <img width="1180" height="805" alt="image" src="https://github.com/user-attachments/assets/350bf6b4-65ce-4cc3-9115-9110ed9450b3" />  

</br></br>

* (**Optional**) To further ensure users are redirected straight to the login, I enabled this option.

  <img width="1885" height="712" alt="image" src="https://github.com/user-attachments/assets/f2aaecbb-1761-4937-b177-63ee04432f33" />  

</br></br>

For Azure, I set it up as follows:
* The **TenantId** must be the same as the one contained in the endpoint.

  <img width="1889" height="897" alt="image" src="https://github.com/user-attachments/assets/b048af8b-5837-4866-906d-f1d040c751e2" />  

</br></br>

* Register the URL ending with **callback** as a **Single-page application** (see image 2).
* Add the **logout URL** (it’s the same as the access URL) – (see image 2).
* Follow the other marked options.

  <img width="844" height="795" alt="image" src="https://github.com/user-attachments/assets/f32188c0-157c-4b1c-b3e8-e04bedb34f7f" />  

</br></br>

* Adjust the API permissions (not sure if all of them are really necessary...).

  <img width="1175" height="629" alt="image" src="https://github.com/user-attachments/assets/2a653124-a6ed-46bf-88ea-f78a30268a71" />
